### ISSUE TITLE
Explicit task pin/unpin interface

### ISSUE THREAD

**Issue by kpamnany** [JuliaLang/julia#52108]
We want the ability to pin tasks to threads so that we can use libraries that expect the same thread to call into it (usually because they use TLS). Currently in Julia, there's no explicit interface to do this and `Base.@async` is often used for this purpose. However, that interface carries additional (and problematic) semantics and really should not be used for this purpose (or at all).

Golang has [LockOSThread](https://pkg.go.dev/runtime#LockOSThread) and [UnlockOSThread](https://pkg.go.dev/runtime#UnlockOSThread). We should add a similar interface, e.g. `pintask()` and `unpintask()`.

Ideally, these would *not* use the current `sticky` field in `Task` because that is inextricably tied up with `Base.@async` semantics (e.g. task stickiness can spread to other tasks -- pinning and unpinning should do no such thing). Although it will complicate the implementation to have separate `sticky` and `pinned` options, it will make removing task stickiness altogether much easier.

Opening this issue to solicit opinions from the community for discussion.

**Comment by kpamnany** [comment]
Of course, we cannot remove the `sticky` field until Julia 2.0 as it would be a breaking change. This is why we should keep this pin/unpin capability completely separate and not reuse `sticky`.

**Comment by NHDaly** [comment]
Interestingly, the Golang API provides a few extra aspects that I don't think we actually need...
From the docs you linked:
- The calling goroutine will always execute in that thread, **and no other goroutine will execute in it**, until the calling goroutine has made as many calls to UnlockOSThread as to LockOSThread.
- If the calling goroutine exits without unlocking the thread, **the thread will be terminated.**

I don't think either of those are necessary for replacing the current uses of `@async`. I wonder what the use-cases are in Golang. Do you think we should mimic these aspects of their API as well?

**Comment by kpamnany** [comment]
I agree that neither of those two features is needed. The first could be quite useful in some situations, but also could be abused.
